enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,933 @@
+---
+title: aws-custom-edge-machineset-local-zones
+authors:
+  - "@mtulio"
+reviewers:
+  - "@rvanderp3"
+  - "@faermanj"
+  - "@patrickdillon"
+  - "@elmiko"
+  - "@JoelSpeed"
+  - "@sdodson"
+approvers:
+  - "@patrickdillon"
+  - "@JoelSpeed"
+  - "@sdodson"
+api-approvers:
+  - None
+creation-date: 2022-09-05
+last-updated: 2022-09-05
+tracking-link:
+  - https://issues.redhat.com/browse/RFE-2782
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Allow Customer-Provisioned edge machine pool for AWS Local Zones
+
+## Summary
+
+Edge adoption is increasing and providers are launching many offerings that fill
+that area. AWS launched the [Local Zones](https://aws.amazon.com/about-aws/global-infrastructure/localzones/)
+and the [Wavelength](https://aws.amazon.com/wavelength/features/) zones to extend the
+VPC network to the metropolitan area from big centers, and for Radio Access Network,
+respectively, to allow application developers to deliver low-latency applications
+at scale to the end users.
+
+As the resources in that locations were designed for specific workloads, and
+also have limited and more expensive cloud services, cluster operators should be
+able to create worker nodes on the edge networks on day zero operations.
+
+This enhancement proposes a new machine pool named `edge`, which will create
+Machine Sets for each availability zone on the AWS Local Zones to provision compute
+resources with `NoSchedule` taint.
```

Comment:
Are we sure that we'll always want this?

Reply:
@sdodson let me reference and summarized previously discussions about this topic:

- [here](https://github.com/openshift/enhancements/pull/1232#discussion_r1045918478) I wrote, and I closed, the `Open Question` about this.
- [here](https://github.com/openshift/enhancements/pull/1232#discussion_r979980109) I detailed the reasons to create `NoSchedule`taints for "edge" nodes. TL;DR: users should have more control in which workload will run on the edge as it is more expensive and can impact the latency when serving requests from/to different locations that the edge is deployed (this is the main feature when choosing to deploy those nodes: achieve low latency to their clients), and shouldn't be balanced as a normal nodes in the parent region (regular AZs). [Here](https://github.com/openshift/enhancements/pull/1232#discussion_r1000204815) I am summarizing the simple benchmark of routing requests from/to 'undesired' locations.

Let me know if you have any questions or thoughts in reasons to not use it.

Reply:
Yeah, I saw the rationale that it's more expensive and I assume that's likely to always be the case given cloud economics.

I'm just not sure we wouldn't potentially have a situation where a customer wants to operate with all workload on localzones with no normal zone workers present. Is that a compact cluster with schedulable control plane nodes with an edge pool or would there be scenarios where they prefer to run infrastructure components like Image Registry localized to the edge pool?

Reply:
Certainly, we can have many variants to use the AWS edge computing (and when [I joined sessions](https://cvws.icloud-content.com/S/AcFiavBeyXTI9iglWF30s9_YTU5m/IMG_0106.JPG?o=AtZsXb_vS24yNb6g1BbMbYs87wYtEb4GWl4QFKoHfQJd&v=1&z=https%3A%2F%2Fp42-content.icloud.com%3A443&x=1&a=CAogt9Lz05vQtdQYZ5M4ksMcnPPGdt_neLC42WPkUMOkgqcSaRDTkuuS2DAY06n-l9gwIgEAUgTYTU5maiclXEKoBCQjU1tsIijqHd15FiKXIa3IOxOZcJM1eLITmYbcGnsJ6xJyJ7kZCXVVM-B2XzCGa1RWX6wXI8cKwL0LfPxmdepI8zjI66DFYjWpxw&e=1672940065&r=55be105d-b36f-4b1f-b985-9c49052863d2-46&s=DN1DAAAijKjP8ugbI8wCkI1fhNw) at AWS re:Invent for this topic of what customers are already doing in the edge, it's really very broad), but most of them could require an extra configuration/customization which I think would be good to choose other install methods than the covered in this EP for the installer using IPI, for example:

- Disconnect clusters may require an extra configuration (associate the LZ subnet into one of the parent's zone route tables to use Nat Gateway there), and will not be covered by this EP
- Some customers operating in countries that do not have regular zones, only edge zones (like Buenos Aires/Argentine, recently announced) may want to data sovereign and low latency deploying the entire cluster in those zones. But Control Planes will not be supported in the Local Zones due to the NLB limitation.
- If no zone workers are present in the regular zones, and edge hasn't NoSchedule taints, the default ingress deployment will fail due to the limitation of NLB/CLB, failing the Console too, and other apps required by it.
- I think they can choose to deploy image-registry in the edge, for example serving the apps there and, maybe their on-prem infra when they have Direct Connect with their infra close to the edge location, but I think it's very specific and can be achieved by [post-installation tasks, as usually do with infra nodes](https://docs.openshift.com/container-platform/4.11/post_installation_configuration/cluster-tasks.html#post-install-creating-infrastructure-machinesets-production)
- Deploying cluster image-registry to the edge, operating in more than one zone groups could imply in latency much higher when serving images inter-edge zones.

So, it seems to be hard to keep a supported cluster in Local Zones without, at least, infra and control planes in the parent region.

Open to hear what do you think, if we may need to cover more scenarios, etc